swagger_zipkin/otel_decorator.py

```diff
@@ -0,0 +1,154 @@
+from opentelemetry.trace.span import format_span_id
+from opentelemetry.trace.span import format_trace_id
+from opentelemetry.trace.span import TraceFlags
+from typing_extensions import ParamSpec
+
+from swagger_zipkin.decorate_client import Client
+from swagger_zipkin.decorate_client import decorate_client
+from swagger_zipkin.decorate_client import Resource
+
+T = TypeVar('T', covariant=True)
+P = ParamSpec('P')
+
+tracer = trace.get_tracer("otel_decorator")
+
+
+class OtelResourceDecorator:
+    """A wrapper to the swagger resource.
+
+    :param resource: A resource object. eg. `client.pet`, `client.store`.
+    :type resource: :class:`swaggerpy.client.Resource` or :class:`bravado_core.resource.Resource`
+    """
+
+    def __init__(self, resource: Resource, client_identifier: str, smartstack_namespace: str) -> None:
+        self.resource = resource
+        self.client_identifier = client_identifier
+        self.smartstack_namespace = smartstack_namespace
+
+    def __getattr__(self, name: str) -> Resource:
+        return decorate_client(self.resource, self.with_headers, name)
+
+    def with_headers(self, call_name: str, *args: Any, **kwargs: Any) -> Any:
+
+        with self.handle_exception():
+            kwargs.setdefault('_request_options', {})
+            request_options: dict = kwargs['_request_options']
+            request_options.setdefault('headers', {})
+
+            operation = getattr(self.resource, call_name)
+            request = construct_request(operation, request_options, **kwargs)  # type: ignore
+
+            url = getattr(request, "url", "")
+            path = getattr(request, "path", "")
+            method = getattr(request, "method", "")
+
+            parent_span = trace.get_current_span()
+            span_name = f"{method} {path}"
+
+            with tracer.start_as_current_span(
+                span_name, kind=trace.SpanKind.CLIENT
+            ) as span:
+                span.set_attribute("url.path", url)
+                span.set_attribute("http.request.method", method)
+                span.set_attribute("client.namespace", self.client_identifier)
+                span.set_attribute("peer.service", self.smartstack_namespace)
+                span.set_attribute("server.namespace", self.smartstack_namespace)
+                span.set_attribute("http.response.status_code", "200")
+
+                self.inject_otel_headers(kwargs, span)
+                self.inject_zipkin_headers(kwargs, span, parent_span)
+
+                try:
+                    operation(*args, **kwargs)
+                except HTTPError as e:
+                    span.set_attribute("error.type", e.__class__.__name__)
+                    span.set_status(trace.Status(trace.StatusCode.ERROR, e.message))
+                    span.set_attribute("http.response.status_code", e.status_code)
+                    raise e
+
+                return operation
```

Comment:
This logic won't work as is. Calling an operation will never raise an `HTTPError` as you must call `result` or `response` on it. You'll need to create a wrapper future class in order to implement the logic you want.

Reply:
Sorry this is an issue

Reply:
Thank you for catching this! Will work on a fix.